src/middleware/immer.ts

```diff
@@ -0,0 +1,66 @@
+import { produce } from 'immer'
+import type { Draft } from 'immer'
+import { State, StateCreator, StoreMutatorIdentifier } from '../vanilla'
+
+type Immer = <
+  T extends State,
+  Mps extends [StoreMutatorIdentifier, unknown][] = [],
+  Mcs extends [StoreMutatorIdentifier, unknown][] = []
+>(
+  initializer: StateCreator<T, [...Mps, ['zustand/immer', never]], Mcs>
+) => StateCreator<T, Mps, [['zustand/immer', never], ...Mcs]>
+
+declare module 'zustand' {
+  // eslint-disable-next-line @typescript-eslint/no-unused-vars
+  interface StoreMutators<S, A> {
+    ['zustand/immer']: WithImmer<S>
+  }
+}
+
+type Write<T extends object, U extends object> = Omit<T, keyof U> & U
+
+type WithImmer<S> = S extends {
+  getState: () => infer T
+  setState: infer SetState
+}
+  ? Write<
+      S,
+      {
+        setState: SetState extends (
+          ...a: [infer _, infer __, ...infer A]
+        ) => infer Sr
+          ? <Nt extends R extends true ? T : Partial<T>, R extends boolean>(
+              nextStateOrUpdater: Nt | ((state: Draft<T>) => void),
+              shouldReplace?: R,
+              ...a: A
+            ) => Sr
+          : never
+      }
+    >
+  : never
+
+type PopArgument<T extends (...a: never[]) => unknown> = T extends (
+  ...a: [...infer A, infer _]
+) => infer R
+  ? (...a: A) => R
+  : never
+
+type ImmerImpl = <T extends State>(
+  storeInitializer: PopArgument<StateCreator<T, [], []>>
+) => PopArgument<StateCreator<T, [], []>>
+
+const immerImpl: ImmerImpl = (initializer) => (set, get, store) => {
+  type T = ReturnType<typeof initializer>
+  const immerSetState: typeof set = (updater, replace) => {
+    const nextState = (
+      typeof updater === 'function' ? produce(updater as any) : updater
+    ) as ((s: T) => T) | T | Partial<T>
+
+    return set(nextState as any, replace)
+  }
+  const s = initializer(immerSetState, get, store)
```

Comment:
It's important to call `initializer` before we mutate `store` so that the mutations that are innermost happen first.

Reply:
Maybe we should add it in the code comment?

Reply:
I realized I got it backwards, the mutation should happen first (as it happens first with all other middlewares). Now the code is more obvious than before, so now should I still put a comment like the following or it's okay to skip it?

```ts
store.setState = (updater, replace) => {
  // ...
}
// We mutate the store first before calling `initializer`
// as we want the inner ("child") mutations to happen later.
// This is also indicated in the type, `immer` returns
// `StateCreator<T, Mps, [['zustand/immer', never], ...Mcs]>`,
// ie `...Mcs` (child mutators) is after `['zustand/immer', never]`
// This is true for all Zustand middlewares btw.

return initializer(store.setState, get, store)
```